src/main/java/com/google/devtools/build/lib/skyframe/CompletionFunction.java

```diff
@@ -303,6 +305,28 @@ public SkyValue compute(SkyKey skyKey, Environment env)
             skyframeActionExecutor.getExecRoot(),
             workspaceNameValue.getName());
 
+    ImmutableMap<String, ActionInput> knownLostOutputs = ImmutableMap.of();
+    try {
+      ensureToplevelArtifacts(env, importantArtifacts, inputMap);
```

Comment:
Can the whole `ensureTopLevelArtifacts` functionality be offloaded to an `ImportantOutputHandler`? It looks an awful lot like what the blaze `ImportantOutputHandler` does, and it would clean up the interface awkwardness introduced by passing in known lost inputs.

Reply:
Thanks for the suggestion, this cleaned up the change quite a bit. I do still need to pass in a new argument (a way to get the generating action for an artifact).